### PR TITLE
v1.9 backports 2021-04-13

### PR DESCRIPTION
**Pull request by jibi**
* #15540 -- bpf: Extend datapath options for K8sVerifier test (@pchaigno)
   * Dropped d380acf ("bpf/Makefile: Disable custom calls on 4.19")
   * Removed `ENABLE_CUSTOM_CALLS`, `ENABLE_DSR_ICMP_ERRORS` and `ENABLE_EGRESS_GATEWAY` references from 0ad05db ("bpf/Makefile: Move global config options to MAX_BASE_OPTIONS") and 68ba5d4 ("bpf/Makefile: Disable Maglev for bpf_host and bpf_lxc")
 * #15590 -- ctmap: do not call InitMapInfo() in init() (@kkourt)
 * #15622 -- Fix ethtool issues (@tklauser)
 * #15260 -- mtu: Switch to v/netlink for querying netdevs (@brb)
 * #15589 -- Kata: Mention incompatibility with host-reachable services or strict KPR in documentation (@qmonnet)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 15540 15590 15622 15260 15589; do contrib/backporting/set-labels.py $pr done 1.9; done
```

### CODE REVIEW

**Comment by jibi** [comment]
test-backport-1.9

File: bpf/Makefile [hunk 1]
```diff
@@ -91,10 +91,13 @@ MAX_BASE_OPTIONS += -DHAVE_LPM_TRIE_MAP_TYPE=1 -DHAVE_LRU_HASH_MAP_TYPE=1 \
...
 	-DENABLE_HOSTPORT=1 -DENABLE_LOADBALANCER -DENABLE_DSR_HYBRID=1 \
-	-DENABLE_DSR_ICMP_ERRORS=1 -DENABLE_IPV4_FRAGMENTS=1 -DENABLE_CUSTOM_CALLS=1
+	-DENABLE_DSR_ICMP_ERRORS=1 -DENABLE_IPV4_FRAGMENTS=1
 ifeq ("$(KERNEL)","netnext")
+# We need to define ENABLE_HOST_SERVICES_FULL to work around complexity issue #14234.
+# We define ENABLE_CUSTOM_CALLS only for net-next (vs. net-next and 4.19) to work
+# around program size issue #15539.
 MAX_BASE_OPTIONS += -DENABLE_TPROXY=1 -DENABLE_REDIRECT_FAST=1 -DENABLE_BANDWIDTH_MANAGER=1 \
-	-DENABLE_HOST_SERVICES_FULL=1 # Required to work around complexity issue #14234.
+	-DENABLE_HOST_SERVICES_FULL=1 -DENABLE_CUSTOM_CALLS=1
```
Comment (pchaigno) [inline bpf/Makefile]: `ENABLE_CUSTOM_CALLS` doesn't exist on 1.9, it's a 1.10 feature. We can drop commit `bpf/Makefile: Disable custom calls on 4.19`.

File: bpf/Makefile [hunk 2]
```diff
@@ -81,13 +81,18 @@ MAX_BASE_OPTIONS = -DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 \
...
+	-DMONITOR_AGGREGATION=3 -DCT_REPORT_FLAGS=0x0002 -DENABLE_HOST_FIREWALL=1
 ifeq ("$(KERNEL)","49")
 # IPSec is incompatible with BPF NodePort so we only enable on 4.9.
 MAX_BASE_OPTIONS += -DENABLE_IPSEC=1 -DNEEDS_RELAX_VERIFIER=1
 else
 MAX_BASE_OPTIONS += -DHAVE_LPM_TRIE_MAP_TYPE=1 -DHAVE_LRU_HASH_MAP_TYPE=1 \
-	-DENABLE_MASQUERADE=1
+	-DENABLE_MASQUERADE=1 -DENABLE_SRC_RANGE_CHECK=1 -DENABLE_NODEPORT=1 \
+	-DENABLE_EXTERNAL_IP=1 -DENABLE_NODEPORT_ACCELERATION=1 \
+	-DENABLE_SESSION_AFFINITY=1 -DENABLE_DSR_ICMP_ERRORS=1 -DENABLE_DSR=1 \
```
Comment (pchaigno) [inline bpf/Makefile]: I think `ENABLE_DSR_ICMP_ERRORS` is a v1.10 feature.

File: bpf/Makefile [hunk 3]
```diff
@@ -81,13 +81,18 @@ MAX_BASE_OPTIONS = -DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 \
...
+	-DENABLE_HOSTPORT=1 -DENABLE_LOADBALANCER -DENABLE_DSR_HYBRID=1 \
+	-DENABLE_DSR_ICMP_ERRORS=1 -DENABLE_IPV4_FRAGMENTS=1 -DENABLE_CUSTOM_CALLS=1 \
```
Comment (pchaigno) [inline bpf/Makefile]: Ditto for `ENABLE_CUSTOM_CALLS`.

File: bpf/Makefile [hunk 4]
```diff
@@ -149,11 +150,9 @@ HOST_OPTIONS = $(LXC_OPTIONS) \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_HOST_FIREWALL: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DPOLICY_VERDICT_NOTIFY:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_NODEPORT_ACCELERATION:-DENABLE_EXTERNAL_IP:-DENABLE_DSR:-DENABLE_DSR_HYBRID:-DENABLE_PREFILTER:-DENABLE_SESSION_AFFINITY:-DENABLE_HOST_FIREWALL:
 
-MAX_HOST_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1 -DENABLE_HOST_FIREWALL=1
+MAX_HOST_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1
 ifneq ("$(KERNEL)","49")
-MAX_HOST_OPTIONS += -DENABLE_HOST_SERVICES_UDP=1 -DENABLE_HOST_SERVICES_TCP=1 \
-	-DENABLE_NODEPORT=1 -DENABLE_NODEPORT_ACCELERATION=1 -DENABLE_EXTERNAL_IP=1 -DENABLE_DSR=1 \
-	-DENABLE_DSR_HYBRID=1 -DENABLE_SESSION_AFFINITY=1
+MAX_HOST_OPTIONS += -DENABLE_EGRESS_GATEWAY=1
```
Comment (pchaigno) [inline bpf/Makefile]: `ENABLE_EGRESS_GATEWAY` is a v1.10 feature.

File: bpf/Makefile [hunk 5]
```diff
@@ -234,8 +230,7 @@ LXC_OPTIONS = \
 # tested as well.
 MAX_LXC_OPTIONS = $(MAX_BASE_OPTIONS) -DENCAP_IFINDEX=1
 ifneq ("$(KERNEL)","49")
-MAX_LXC_OPTIONS += -DENABLE_IPV4_FRAGMENTS=1 -DENABLE_NODEPORT=1 -DENABLE_SESSION_AFFINITY=1 \
-	-DENABLE_EXTERNAL_IP=1 -DENABLE_HOSTPORT=1 -DENABLE_LOADBALANCER=1
+MAX_LXC_OPTIONS += -DENABLE_EGRESS_GATEWAY=1
```
Comment (pchaigno) [inline bpf/Makefile]: Ditto.

File: bpf/Makefile (hunk 4 above)
Comment (pchaigno) [reply bpf/Makefile]: The commit message probably also needs to be updated accordingly.

**Review by pchaigno — CHANGES_REQUESTED** [review]
A couple changes are required for my PR. If you prefer to drop it, it's also fine; I can take care of that backport.

**Review by tklauser — APPROVED** [review]
> #15622 -- Fix ethtool issues (@tklauser)

LGTM for this one, thanks 🚀

File: bpf/Makefile (hunk 4 above)
Comment (jibi) [reply bpf/Makefile]: :man_facepalming:

**Review by qmonnet — APPROVED** [review]
> #15589 -- Kata: Mention incompatibility with host-reachable services or strict KPR in documentation (@qmonnet)

Looks good, thank you.

**Comment by jibi** [comment]
Compilation fails with
```
# github.com/cilium/cilium/pkg/datapath/linux/ethtool
../pkg/datapath/linux/ethtool/ethtool.go:66:8: undefined: unix.ETHTOOL_GDRVINFO
```

I may just need to update `golang.org/x/sys/unix`

**Review by pchaigno — APPROVED** [review]
My commits look good :+1: Thanks @jibi!

**Comment by jibi** [comment]
test-backport-1.9

**Comment by jibi** [comment]
test-backport-1.9

**Comment by jibi** [comment]
retest-gke

**Comment by jibi** [comment]
retest-gke

**Review by brb — APPROVED** [review]
My changes LGTM, thanks!

**Comment by jibi** [comment]
Complexity issue on 4.19, looking into it
```
BPF program is too large. Processed 131073 insn
```

**Comment by kkourt** [comment]
My commit looks good! Thanks!

**Review by kkourt — APPROVED** [review]
My commit looks good! thanks!

**Comment by jibi** [comment]
test-backport-1.9

**Pull request merged**